File: lua/game.lua
local Player = {}
Player.__index = Player

function Player:new(name)
    local self = setmetatable({}, Player)
    self.name = name or "Hero"
    self.health = 100
    self.score = 0
    self.level = 1
    self.x = 0
    self.y = 0
    return self
end

function Player:move(dx, dy)
    self.x = self.x + dx
    self.y = self.y + dy
end

function Player:take_damage(amount)
    self.health = math.max(0, self.health - amount)
    return self.health > 0
end

function Player:add_score(points)
    self.score = self.score + points
end

function Player:heal()
    if self.health < 100 then
        local heal_amount = math.min(20, 100 - self.health)
        self.health = self.health + heal_amount
        print(string.format("Healed for %d HP", heal_amount))
    else
        print("Already at full health!")
    end
end

function Player:display_status()
    print(string.format("\n[%s] HP: %d | Score: %d | Level: %d", self.name, self.health, self.score, self.level))
    print(string.format("Position: (%d, %d)", self.x, self.y))
end

local GameEngine = {}
GameEngine.__index = GameEngine

function GameEngine:new(player)
    local self = setmetatable({}, GameEngine)
    self.player = player
    self.running = true
    return self
end

function GameEngine:run()
    print("\n=== Lua Game Engine ===")
    print("Welcome to the Lua Game!\n")

    while self.running and self.player.health > 0 do
        self.player:display_status()
        self:handle_input()
        self:update_game()
    end

    self:end_game()
end

function GameEngine:handle_input()
    io.write("Commands: (m)ove, (a)ttack, (h)eal, (q)uit: ")
    local input = io.read():lower()

    if input:len() > 0 then
        local command = input:sub(1, 1)
        if command == "m" then
            local dx = math.random(-2, 2)
            local dy = math.random(-2, 2)
            self.player:move(dx, dy)
            print(string.format("Moved to (%d, %d)", self.player.x, self.player.y))
        elseif command == "a" then
            self:attack()
        elseif command == "h" then
            self.player:heal()
        elseif command == "q" then
            self.running = false
        else
            print("Unknown command")
        end
    end
end

function GameEngine:attack()
    local damage = math.random(10, 30)
    print(string.format("You attacked for %d damage!", damage))
    self.player:add_score(damage)
end

function GameEngine:update_game()
    if math.random() < 0.15 then
        local damage = math.random(5, 15)
        print(string.format("\nEnemy attacked for %d damage!", damage))
        self.player:take_damage(damage)
    end
end

function GameEngine:end_game()
    print("\n=== GAME OVER ===")
    print(string.format("Final Score: %d", self.player.score))
end

-- Main
math.randomseed(os.time())
io.write("Enter your player name: ")
local name = io.read()
local player = Player:new(name)
local engine = GameEngine:new(player)
engine:run()

File: swift/game.swift
import Foundation

struct Player {
    var name: String
    var health: Int = 100
    var score: Int = 0
    var level: Int = 1
    var x: Int = 0
    var y: Int = 0

    mutating func move(dx: Int, dy: Int) {
        x += dx
        y += dy
    }

    mutating func takeDamage(_ amount: Int) -> Bool {
        health = max(0, health - amount)
        return health > 0
    }

    mutating func addScore(_ points: Int) {
        score += points
    }

    mutating func heal() {
        if health < 100 {
            let healAmount = min(20, 100 - health)
            health += healAmount
            print("Healed for \(healAmount) HP")
        } else {
            print("Already at full health!")
        }
    }

    func displayStatus() {
        print("\n[\(name)] HP: \(health) | Score: \(score) | Level: \(level)")
        print("Position: (\(x), \(y))")
    }
}

class GameEngine {
    var player: Player
    var running: Bool = true

    init(player: Player) {
        self.player = player
    }

    func run() {
        print("\n=== Swift Game Engine ===")
        print("Welcome to the Swift Game!\n")

        while running && player.health > 0 {
            player.displayStatus()
            handleInput()
            updateGame()
        }

        endGame()
    }

    private func handleInput() {
        print("Commands: (m)ove, (a)ttack, (h)eal, (q)uit: ", terminator: "")
        fflush(stdout)

        if let input = readLine()?.lowercased(), let command = input.first {
            switch command {
            case "m":
                let dx = Int.random(in: -2...2)
                let dy = Int.random(in: -2...2)
                player.move(dx: dx, dy: dy)
                print("Moved to (\(player.x), \(player.y))")
            case "a":
                attack()
            case "h":
                player.heal()
            case "q":
                running = false
            default:
                print("Unknown command")
            }
        }
    }

    private func attack() {
        let damage = Int.random(in: 10...30)
        print("You attacked for \(damage) damage!")
        player.addScore(damage)
    }

    private func updateGame() {
        if Double.random(in: 0.0..<1.0) < 0.15 {
            let damage = Int.random(in: 5...15)
            print("\nEnemy attacked for \(damage) damage!")
            _ = player.takeDamage(damage)
        }
    }

    private func endGame() {
        print("\n=== GAME OVER ===")
        print("Final Score: \(player.score)")
    }
}

let playerName = {
    print("Enter your player name: ", terminator: "")
    fflush(stdout)
    return readLine() ?? ""
}()

var player = Player(name: playerName.isEmpty ? "Hero" : playerName)
let engine = GameEngine(player: player)
engine.run()
